File: common/src/main/java/com/terrasect/common/handler/ClimateHandler.java
package com.terrasect.common.handler;

import com.terrasect.common.Context;
import com.terrasect.common.generation.World;
import com.terrasect.common.helpers.ClimateModifier;
import net.minecraft.world.level.biome.Climate;

public final class ClimateHandler {
    private ClimateHandler() {
    }

    public static Climate.TargetPoint modifyTargetPoint(
            Context context, int x, int y, int z, Climate.TargetPoint original) {

        if (context == null || original == null) {
            return original;
        }

        var blockX = x << 2;
        var blockZ = z << 2;

        var traversal = World.traverse(context, blockX, blockZ);
        if (traversal == null || traversal.region == null) {
            return original;
        }

        var climate = traversal.region.definition().climate();
        if (climate == null || !climate.hasRanges()) {
            return original;
        }

        var edgeFactor = 1.0f - traversal.edgeDistance;
        return ClimateModifier.modify(climate, original, edgeFactor);
    }
}

File: common/src/main/kotlin/terrasect/handler/ClimateHandler.kt
package terrasect.handler

import net.minecraft.world.level.biome.Climate
import terrasect.extender.ClimateTargetPointExtender
import terrasect.extender.NoiseChunkExtender
import terrasect.generation.ChunkContext
import terrasect.instrumentation.TerrasectInstr
import terrasect.instrumentation.TerrasectMetricEvent

private val instr = TerrasectInstr.climate

object ClimateHandler {
  fun modifyClimate(
    quadX: Int,
    quadY: Int,
    quadZ: Int,
    climate: Climate.TargetPoint,
    noiseChunk: NoiseChunkExtender?,
  ) {
    val blockX = quadX shl 2
    val blockZ = quadZ shl 2
    val chunk: ChunkContext? = noiseChunk?.`terrasect$getChunk`()?.`terrasect$getContext`()

    if (chunk == null) {
      instr.count(TerrasectMetricEvent.CLIMATE_CHUNK_MISSING)
      return
    }

    val region = chunk.getRegion(blockX, blockZ)
    if (region == null) {
      return
    }

    val constraints = region.climate
    @Suppress("CAST_NEVER_SUCCEEDS") val extender = climate as ClimateTargetPointExtender
    if (constraints != null) {
      constraints.temperature?.let { range ->
        extender.`terrasect$setTemperature`(climate.temperature.coerceIn(range.min, range.max))
      }
      constraints.humidity?.let { range ->
        extender.`terrasect$setHumidity`(climate.humidity.coerceIn(range.min, range.max))
      }
      constraints.continentalness?.let { range ->
        extender.`terrasect$setContinentalness`(
          climate.continentalness.coerceIn(range.min, range.max)
        )
      }
      constraints.erosion?.let { range ->
        extender.`terrasect$setErosion`(climate.erosion.coerceIn(range.min, range.max))
      }
      constraints.depth?.let { range ->
        extender.`terrasect$setDepth`(climate.depth.coerceIn(range.min, range.max))
      }
      constraints.weirdness?.let { range ->
        extender.`terrasect$setWeirdness`(climate.weirdness.coerceIn(range.min, range.max))
      }
      instr.count(TerrasectMetricEvent.CLIMATE_APPLIED)
    }
  }
}
